lib/datadog/tracing/context_provider.rb

```diff
@@ -61,6 +62,18 @@ def local=(ctx)
       def local(storage = Thread.current)
         storage[@key] ||= Context.new
       end
+
+      # Ensure two instances of {FiberLocalContext} do not conflict.
+      # We previously used {FiberLocalContext#object_id} to ensure uniqueness
+      # but the VM is allowed to reuse `object_id`, allow for the possibility that
+      # a new FiberLocalContext would be able to read an old FiberLocalContext's
+      # value.
+      @mutex = Mutex.new
+      @unique_instance_id = Datadog::Core::Utils::Sequence.new
+
+      def self.next_instance_id
+        @mutex.synchronize { @unique_instance_id.next }
+      end
```

Comment:
Minor: I usually favor putting these into constants, to make it even more clear they are not instance-scoped, but the effect is exactly the same so it's really just for readibility.

```suggestion
      UNIQUE_INSTANCE_MUTEX = Mutex.new
      UNIQUE_INSTANCE_GENERATOR = Datadog::Core::Utils::Sequence.new

      def self.next_instance_id
        UNIQUE_INSTANCE_MUTEX.synchronize { UNIQUE_INSTANCE_GENERATOR.next }
      end
```

(Potentially add a `private_constant :UNIQUE_INSTANCE_MUTEX` + generator as well maybe?)

Reply:
I don't really mind either way, but I hate to disappoint a reviewer :)